cumulus/client/bootnodes/src/discovery.rs

```diff
@@ -0,0 +1,460 @@
+	relay_chain_interface: Arc<dyn RelayChainInterface>,
+	relay_chain_network: Arc<dyn NetworkService>,
+	latest_relay_chain_hash: Option<Hash>,
+	key_being_discovered: Option<KademliaKey>,
+	paranode_protocol_name: ProtocolName,
+	pending_responses: FuturesUnordered<
+		BoxFuture<
+			'static,
+			(PeerId, Result<Result<(Vec<u8>, ProtocolName), RequestFailure>, oneshot::Canceled>),
+		>,
+	>,
+	direct_requests: HashSet<PeerId>,
+	find_node_queries: HashSet<PeerId>,
+	attempts_left: u32,
+	succeeded: bool,
+}
+
+impl BootnodeDiscovery {
+	/// Create a new bootnode discovery service.
+	pub fn new(
+		BootnodeDiscoveryParams {
+			para_id,
+			parachain_network,
+			parachain_genesis_hash,
+			parachain_fork_id,
+			relay_chain_interface,
+			relay_chain_network,
+			paranode_protocol_name,
+		}: BootnodeDiscoveryParams,
+	) -> Self {
+		Self {
+			para_id_scale_compact: CompactRef(&para_id).encode(),
+			parachain_network,
+			parachain_genesis_hash,
+			parachain_fork_id,
+			relay_chain_interface,
+			relay_chain_network,
+			latest_relay_chain_hash: None,
+			key_being_discovered: None,
+			paranode_protocol_name,
+			pending_responses: FuturesUnordered::default(),
+			direct_requests: HashSet::new(),
+			find_node_queries: HashSet::new(),
+			attempts_left: MAX_DISCOVERY_ATTEMPTS,
+			succeeded: false,
+		}
+	}
+
+	async fn current_epoch(&mut self, hash: Hash) -> RelayChainResult<Epoch> {
+		let res = self
+			.relay_chain_interface
+			.call_runtime_api("BabeApi_current_epoch", hash, &[])
+			.await?;
+		Decode::decode(&mut &*res).map_err(Into::into)
+	}
+
+	fn epoch_key(&self, randomness: Randomness) -> KademliaKey {
+		self.para_id_scale_compact
+			.clone()
+			.into_iter()
+			.chain(randomness.into_iter())
+			.collect::<Vec<_>>()
+			.into()
+	}
+
+	/// Start bootnode discovery.
+	async fn start_discovery(&mut self) -> RelayChainResult<()> {
+		let Some(hash) = self.latest_relay_chain_hash else {
+			error!(
+				target: LOG_TARGET,
+				"Failed to start bootnode discovery: no relay chain hash available. This is a bug.",
+			);
+			// This is a graceful panic via the failure of essential task.
+			return Err(RelayChainError::GenericError("no relay chain hash available".to_string()));
+		};
+
+		let current_epoch = self.current_epoch(hash).await?;
+		let current_epoch_key = self.epoch_key(current_epoch.randomness);
+		self.key_being_discovered = Some(current_epoch_key.clone());
+		self.relay_chain_network.get_providers(current_epoch_key.clone());
+
+		debug!(
+			target: LOG_TARGET,
+			"Started discovery of parachain bootnode providers for current epoch key {}",
+			hex::encode(current_epoch_key),
+		);
+
+		Ok(())
+	}
+
+	/// Start bootnode discovery if needed. Returns `false` if the discovery event loop should be
+	/// terminated.
+	async fn maybe_start_discovery(&mut self) -> RelayChainResult<bool> {
+		// Start discovery if it is not currently in progress.
+		if self.key_being_discovered.is_none() &&
```

Comment:
nit: I usually prefer early returns, feel free to ignore

```rust
if key_being_disocvered.is_some() || pending .. {
    return Ok(true)
}

if succeeded {
    return Ok(false)
}

if attempts > 0 {
    return Ok(true)
}
...
```

Reply:
Me too, but it is too C++-ish :D